### PR TITLE
Elaborate on optional chaining short-circuiting

### LINKED ISSUE #17099: Optional Chaining page should explain the chaining aspect
### MDN URL

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining

### What specific section or headline is this issue about?

Most of the doc, but "Syntax" is crucial starting point

### What information was incorrect, unhelpful, or incomplete?

The current page presents `?.` as an operator similar to `.` that takes a single `?.prop` / `?.[expr]` / `?.(args)` on the right side, and may safely return `undefined` instead of error.

Under that model, a long chain of `?.` followed by `.` sounds like separate operators, and if the first `?.` returns undefined the latter `.` should error.  But that's not how it works!

```
> undefined?.split(',').slice().map().reduce()
undefined
```
Contrast this with forcing evaluation as separate operators:
```
> t = (undefined)?.split(',').slice()
undefined
> t.map().reduce()
Uncaught TypeError: Cannot read properties of undefined (reading 'map')

> ((((undefined)?.split(',')).slice()).map()).reduce()
Uncaught TypeError: Cannot read properties of undefined (reading 'slice')
```

IIUC the ECMA spec syntax productions, `?.` actually begins an `OptionalChain`, with following `.prop`, `.method()` etc. all being part of it:
```
OptionalChain :
    ?. Arguments
    ?. [ Expression ]
    ?. IdentifierName
    ?. TemplateLiteral
    ?. PrivateIdentifier
    OptionalChain Arguments
    OptionalChain [ Expression ]
    OptionalChain . IdentifierName
    OptionalChain TemplateLiteral
    OptionalChain . PrivateIdentifier
```
If the left side is null/undefined, _the whole chain is skipped_!

### What did you expect to see?

- [ ] Syntax section covering the concept OptionalChain as a _syntactic_ construct.
    Without understanding this parsing, you can't form a correct model of what it does.

- [ ] Take care with misleading word "operator".

- [ ] Examples of longer chains and explanation that the whole chain is skipped.

    - [ ] Examples contrasting `foo?.bar.baz` vs `foo?.bar?.baz` and explanation why you might want one of the other.
    - [ ] Explanation that `foo?.bar?.baz` does parse as 2 chains `(foo?.bar)?.baz`.

- [ ] Can a chain be explained by rewriting to equivalent logic?
    Something like `isNullish(foo) ? undefined : foo.bar.baz`?  If yes, I think this'd be helpful to understanding.

### Do you have any supporting links, references, or citations?

the spec already linked on the page, https://tc39.es/ecma262/multipage/ecmascript-language-expressions.html#prod-OptionalExpression

### Do you have anything more you want to share?

_No response_

### MDN metadata

<details>
<summary>Page report details</summary>

* Folder: `en-us/web/javascript/reference/operators/optional_chaining`
* MDN URL: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining
* GitHub URL: https://github.com/mdn/content/blob/main/files/en-us/web/javascript/reference/operators/optional_chaining/index.md
* Last commit: https://github.com/mdn/content/commit/d83d0a726583cb808beade63bae9c173c43144f6
* Document last modified: 2022-04-26T02:34:14.000Z

</details>

### PR DESCRIPTION
**Pull request by Josh-Cena**
#### Summary

Fix #17099.

This is too hard to get right 😅 @cben would be helpful if you can check if this describes what you have in mind.

#### Metadata

- [ ] Adds a new document
- [ ] Rewrites (or significantly expands) a document
- [x] Fixes a typo, bug, or other error

### CODE REVIEW

File: files/en-us/web/javascript/reference/operators/optional_chaining/index.md [hunk 1]
```diff
@@ -165,9 +132,55 @@ You can use [bracket notation](/en-US/docs/Web/JavaScript/Reference/Operators/Pr
 
 ```js
 const arr = ['a', 'b', 'c', 'd']
-let arrayItem = arr?.[42];
+const arrayItem = arr?.[42];
 ```
 
+### Short-circuiting
```
Comment (teoli2003) [inline files/en-us/web/javascript/reference/operators/optional_chaining/index.md:138]: I found this easy to understand.

File: files/en-us/web/javascript/reference/operators/optional_chaining/index.md [hunk 2]
```diff
@@ -109,53 +108,21 @@ let result = someInterface.customMethod?.();
...
-```
-
 ### Optional chaining with expressions
 
 You can also use the optional chaining operator when accessing properties with an expression using
 [the bracket notation of the property accessor](/en-US/docs/Web/JavaScript/Reference/Operators/Property_Accessors#bracket_notation):
 
 ```js
-let nestedProp = obj?.['prop' + 'Name'];
+const nestedProp = obj?.['prop' + 'Name'];
```
Comment (teoli2003) [inline files/en-us/web/javascript/reference/operators/optional_chaining/index.md:117]: Not a problem, but why the `+`?

**Review by teoli2003 — COMMENTED** [review]
Re-request my review when you had the feedback requested.

File: files/en-us/web/javascript/reference/operators/optional_chaining/index.md (hunk 2 above)
Comment (Josh-Cena) [reply files/en-us/web/javascript/reference/operators/optional_chaining/index.md:117]: To demonstrate you can have arbitrary expressions in the brackets, I think

File: files/en-us/web/javascript/reference/operators/optional_chaining/index.md (hunk 1 above)
Comment (Josh-Cena) [reply files/en-us/web/javascript/reference/operators/optional_chaining/index.md:138]: It's not particularly hard to understand as areader, but our discussion of #17099 has covered so many nuances that I'm not totally satisfied with this as a writer :P

File: files/en-us/web/javascript/reference/operators/optional_chaining/index.md [hunk 3]
```diff
@@ -38,7 +38,6 @@ Optional chaining cannot be used on a non-declared root object, but can be used
 ```js
 obj.val?.prop
 obj.val?.[expr]
-obj.arr?.[index]
```
Comment (cben) [inline files/en-us/web/javascript/reference/operators/optional_chaining/index.md:41]: :+1: _syntactically_ there is no difference between obj[computedProp] and array[computedIndex], and even at run-time I think they follow common path — the [] operator [converts numbers to strings](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Property_Accessors#property_names) and Arrays simply know how to respond to number-like property names.

Comment (cben) [reply files/en-us/web/javascript/reference/operators/optional_chaining/index.md:41]: consider also merging "Array item access with optional chaining" section into "Optional chaining with expressions" section.  They are effectively the same but currently separated by unrelated "not valid on left side".

Comment (Josh-Cena) [reply files/en-us/web/javascript/reference/operators/optional_chaining/index.md:41]: Done.

**Review by teoli2003 — APPROVED** [review]
I think we can merge this. Anything else can go into follow-up PRs.

File: files/en-us/web/javascript/reference/operators/optional_chaining/index.md [hunk 4]
```diff
@@ -109,65 +108,77 @@ let result = someInterface.customMethod?.();
...
-// Using optional chaining with function calls
-function doSomething(onContent, onError) {
-  try {
-   // ... do something with the data
-  }
-  catch (err) {
-    onError?.(err.message); // no exception if onError is undefined
-  }
-}
+const arr = ['a', 'b', 'c', 'd']
```
Comment (cben) [inline files/en-us/web/javascript/reference/operators/optional_chaining/index.md:123]:
```suggestion
const arr = ['a', 'b', 'c', 'd'];
```
(valid either way of course, just distracting if different from surrounding style)

**Comment by cben** [comment]
ah I see you already merged.  This is certainly an improvement :+1:, I still have some comments but I'll try to make a follow PR from them.

**Pull request merged**